IDE/STM32Cube/wolfssl_example.c

```diff
@@ -1468,25 +1469,264 @@ static void PrintMemStats(void)
 #endif
 }
 
-#if 0
-static void* wolfMallocCb(size_t size)
+/* UART DMA IO Routines */
+#ifndef B115200
+#define B115200 115200
+#endif
+
+/* Max buffer for a single TLS frame */
+#ifndef MAX_RECORD_SIZE
+#define MAX_RECORD_SIZE (16 * 1024)
+#endif
+
+typedef struct {
+	uint16_t curr_index;
+        uint16_t data_len;
+	char buf[MAX_RECORD_SIZE];
+} tls13_buf;
+
+/* This sets which UART to do the TLS 1.3 connection over.  It is something you
+ * will have to configure in STMCubeIDE and then change here. */
+#define TLS_UART huart2
+extern UART_HandleTypeDef TLS_UART;
+
+static unsigned int msg_length = 0;
+
+void HAL_UARTEx_RxEventCallback(UART_HandleTypeDef *huart, uint16_t Size) {
+   if (huart->Instance == TLS_UART.Instance) {
+       msg_length = Size;
+   }
+}
+
+static int uartIORx(WOLFSSL *ssl, char *buf, int sz, void *ctx)
 {
-	void* ptr = malloc(size);
-	if (ptr == NULL) {
-		printf("BREAK!\n");
-	}
-	return ptr;
+    HAL_StatusTypeDef status;
+    tls13_buf *tb = ctx;
+
+#ifdef DEBUG_UART_IO
+    printf("UART Read: In %d\n", sz);
+#endif
+
+    if (tb->curr_index + sz <= tb->data_len) {
+        XMEMCPY(buf, tb->buf + tb->curr_index, sz);
+        tb->curr_index += sz;
+#ifdef DEBUG_UART_IO
+        printf("UART Read1: Out %d\n", sz);
+#endif
+        return sz;
+    }
+
+    msg_length = 0;
+    XMEMSET(tb, 0, sizeof(*tb));
+
+    /* Now get the bytes sent */
+    status = HAL_UARTEx_ReceiveToIdle_DMA(&TLS_UART, (uint8_t *)tb->buf, MAX_RECORD_SIZE);
+
+    if (status != HAL_OK) {
+        return WOLFSSL_CBIO_ERR_WANT_READ;
+    } else {
+        while(msg_length == 0) {
```

Comment:
Please add a comment and possibly some timeout logic here...

Reply:
I'd rather not timeout here.  If this is the server, then its waiting here for the client to send client hello, but if the other side is linux then the user will need to run the client app with `sudo` in which case it might take longer and it would be frustrating to timeout while entering the password. Its easy enough to reset the board if they accidentally executed the server.